src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,217 @@
+        }
+        eligible_packets
+    }
+
+    pub fn get_total_privacy_budget(&self) -> u32 {
+        self.total_privacy_budget
+    }
+
+    /// Budget is increased for each key atomically in Redis
+    fn increase_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<bool> {
+        let key = packet.to_string();
+        let (budget_consumed,): (u32,) =
+            transaction(&mut self.redis_connection, &[key.clone()], |con, pipe| {
+                let val: u32 = con.get(&key).unwrap_or(0);
+                // increment
+                pipe.set(&key, val + budget).ignore().get(&key).query(con)
+            })?;
+        if budget_consumed > self.get_total_privacy_budget() {
+            Ok(false)
+        } else {
+            Ok(true)
+        }
+    }
+
+    /// Budget is decreased for each ineligible key in Redis
+    fn decrease_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<()> {
+        self.redis_connection.decr(packet.to_string(), budget)?;
+        Ok(())
+    }
+}
+
+impl From<RedisError> for Error {
+    fn from(_e: RedisError) -> Self {
+        Error::RedisError
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::thread_rng;
+    use std::sync::Mutex;
+
+    lazy_static! {
+        static ref PRIVACY_BUDGET: Mutex<PrivacyBudget> =
+            Mutex::new(PrivacyBudget::new("127.0.0.1", None, false).unwrap());
+    }
+
+    fn create_ciphertexts() -> [BlindedMatchKey; 4] {
+        let mut rng = thread_rng();
+
+        let k1 = DecryptionKey::new(&mut rng);
+        let mk1 = RistrettoPoint::random(&mut rng);
+        let mk2 = RistrettoPoint::random(&mut rng);
+        let mk3 = RistrettoPoint::random(&mut rng);
+        let mk4 = RistrettoPoint::random(&mut rng);
+        let r = Scalar::random(&mut rng);
+
+        let c1 = BlindedMatchKey::new(k1, mk1, r);
+        let c2 = BlindedMatchKey::new(k1, mk2, r);
+        let c3 = BlindedMatchKey::new(k1, mk3, r);
+        let c4 = BlindedMatchKey::new(k1, mk4, r);
+
+        [c1, c2, c3, c4]
+    }
+
+    fn clear_redis_keys(privacy_budget: &mut PrivacyBudget, packets: &[BlindedMatchKey]) {
+        for packet in packets.iter() {
+            #![allow(clippy::let_unit_value)]
+            let _: () = privacy_budget
+                .redis_connection
+                .del(packet.to_string())
+                .expect("Failed to delete keys");
+        }
+    }
+
+    fn create_privacy_budget_ciphertexts(
+        privacy_budget: &mut PrivacyBudget,
+    ) -> [BlindedMatchKey; 4] {
+        let packets = create_ciphertexts();
+        clear_redis_keys(&mut *privacy_budget, &packets);
+        packets
+    }
+
+    #[test]
+    fn privacy_budget_didnot_exceed() {
+        let mut privacy_budget = PRIVACY_BUDGET.lock().unwrap();
+        let packets = create_privacy_budget_ciphertexts(&mut *privacy_budget);
+        let eligible_packets = privacy_budget.get_eligible_packets_with_budget(&packets, 20);
```

Comment:
The 20 here is relative to a constant.

Reply:
20 here is an absolute budget which was passed to be consumed for all the match keys in the query. Is the understanding correct?